haystack/nodes/prompt/prompt_node.py

```diff
@@ -0,0 +1,822 @@
+        :param kwargs: Additional keyword arguments to pass to the invocation layer
+        :return: A list of model generated responses for the prompt or prompts
+        """
+        output = self.model_invocation_layer.invoke(prompt=prompt, **kwargs)
+        return output
+
+    def run(
+        self,
+        query: Optional[str] = None,
+        file_paths: Optional[List[str]] = None,
+        labels: Optional[MultiLabel] = None,
+        documents: Optional[List[Document]] = None,
+        meta: Optional[dict] = None,
+    ) -> Tuple[Dict, str]:
+        raise NotImplementedError("This method should never be implemented in the derived class")
+
+    def run_batch(
+        self,
+        queries: Optional[Union[str, List[str]]] = None,
+        file_paths: Optional[List[str]] = None,
+        labels: Optional[Union[MultiLabel, List[MultiLabel]]] = None,
+        documents: Optional[Union[List[Document], List[List[Document]]]] = None,
+        meta: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]] = None,
+        params: Optional[dict] = None,
+        debug: Optional[bool] = None,
+    ):
+        raise NotImplementedError("This method should never be implemented in the derived class")
+
+
+class PromptNode(BaseComponent):
+    """
+    The PromptNode class is the central abstraction in Haystack's large language model (LLM) support. PromptNode
+    supports multiple NLP tasks out of the box. PromptNode allows users to perform multiple tasks, such as
+    summarization, question answering, question generation etc., using a single, unified model within the Haystack
+    framework.
+
+    One of the benefits of PromptNode is that it allows users to define and add additional prompt templates
+    that the model supports. Defining additional prompt templates enables users to extend the model's capabilities
+    and use it for a broader range of NLP tasks within the Haystack ecosystem. Prompt engineers define templates
+    for each NLP task and register them with PromptNode. The burden of defining templates for each task rests on
+    the prompt engineers, not the users.
+
+    Using an instance of PromptModel class, we can create multiple PromptNodes that share the same model, saving
+    the memory and time required to load the model multiple times.
+
+    PromptNode also supports multiple model invocation layers: Hugging Face transformers and OpenAI with an
+    ability to register additional custom invocation layers.
+
+    """
+
+    outgoing_edges: int = 1
+    prompt_templates: Dict[str, PromptTemplate] = {
+        prompt_template.name: prompt_template for prompt_template in PREDEFINED_PROMPT_TEMPLATES  # type: ignore
+    }
+
+    def __init__(
+        self,
+        model_name_or_path: Union[str, PromptModel] = "google/flan-t5-base",
+        default_prompt_template: Optional[Union[str, PromptTemplate]] = None,
+        output_variable: Optional[str] = None,
+        max_length: Optional[int] = 100,
```

Comment:
If the max_length value is optional I think the default value should be None. What do you think?

If we need the default value to be set then I would recommend removing the Optional type.